webbpsf/optics.py

```diff
@@ -2,17 +2,284 @@
 import os
 import poppy
 import numpy as np
+import scipy
+import matplotlib
 
-import astropy.table
+import astropy.table as table
 import astropy.io.fits as fits
+import astropy.units as units
 
 from . import utils
-from .webbpsf_core import _log
 
+import logging
+_log = logging.getLogger('webbpsf')
+
+
+
+#######  Classes for modeling aspects of JWST's segmented active primary #####
+
+def segment_zernike_basis(segnum=1, nterms=15, npix=512, outside = np.nan):
+    """ Basis set in the style of poppy.zernike.zernike_basis for segment-level
+    Zernike polynomials for one segment at a time in JWST's aperture.
+
+    Parameters
+    ------------
+    segnum : integer
+        1 to 18, number of JWST segment. Uses same numbering convention as the WSS.
+    nterms : integer
+        Number of Zernike polynomial terms to return
+    npix : integer
+        Number of pixels per side of the array
+    outside : float
+        Value to fill the array with outside of the valid segment.
+
+    """
+    from .webbpsf_core import segname
+
+    aper = JWST_primary_aperture(label_segments=True)
+    w = poppy.Wavefront(npix=npix,diam=JWST_primary_aperture.circumscribed_diameter)
+    segmask = aper.get_transmission(w)
+
+    segname = segname(segnum)
+    cenx, ceny = aper.seg_centers[segname]
+
+    seg_radius = 1.517/2 # nominal point to point diam for A and B segments;
+                         # ignoring slight departures from ideal hexes for now.
+
+    y,x = w.coordinates()
+
+    r = np.sqrt( (y-ceny)**2 + (x-cenx)**2)/seg_radius
+    theta =  np.arctan2((y-ceny) / seg_radius, (x-cenx) / seg_radius)
+    r[segmask != segnum] = np.nan
+    theta[segmask != segnum] = np.nan
+
+    wg = np.where(segmask==segnum)
+    outzerns = np.full((nterms, npix,npix), outside, dtype=float)
+    outzerns_tmp = poppy.zernike.zernike_basis(nterms=nterms,rho=r[wg], theta=theta[wg], outside=outside)
+    for iz in range(nterms):
+        outzerns[iz][wg] = outzerns_tmp[iz]
+
+    return outzerns
+
+
+class JWST_primary_aperture(poppy.AnalyticOpticalElement):
```

Comment:
For PEP8 compliance, this should be `JWSTPrimaryAperture` (If you ask some of the LSST people, maybe `JwstPrimaryAperture`... but that looks worse to me.)